src/nvim/os/stream.c

```diff
@@ -0,0 +1,26 @@
+// Functions for working with stdio streams (as opposed to RStream/WStream).
+
+#include <stdio.h>
+#include <stdbool.h>
+
+#include <uv.h>
+
+#ifdef INCLUDE_GENERATED_DECLARATIONS
+# include "os/stream.c.generated.h"
+#endif
+
+/// Sets the stream associated with `fd` to "blocking" mode.
+///
+/// @return `0` on success, or `-errno` on failure.
+int stream_set_blocking(int fd, bool blocking)
+{
+  int flags = fcntl(fd, F_GETFL, 0);
+  int err = 0;
+  if (!blocking && !(flags & O_NONBLOCK)) {
+    err = fcntl(fd, F_SETFL, flags | O_NONBLOCK);
+  } else if (blocking && (flags & O_NONBLOCK)) {
+    err = fcntl(fd, F_SETFL, flags & ~O_NONBLOCK);
```

Comment:
> Here I'm calling fcntl instead of the previous libuv-based implementation (see below for reference), because in the case where some component other than main.c was setting stderr to non-blocking, using the libuv-based implementation would result in this runtime assertion:

Would it help to create a new loop with `uv_loop_init()` and use it instead of `uv_default_loop()`?
```c
int stream_set_blocking(int fd, bool blocking)
{
  uv_pipe_t stream;
  uv_loop_t loop;
  uv_loop_init(&loop);
  uv_pipe_init(&loop, &stream, 0);
  uv_pipe_open(&stream, fd);
  int retval = uv_stream_set_blocking((uv_stream_t *)&stream, blocking);
  uv_close((uv_handle_t *)&stream, NULL);
  uv_run(&loop, UV_RUN_NOWAIT); // do we need to run the loop at least once?
  uv_loop_close(&loop);
  return retval;
}
```